File: RemixedDungeon/src/main/java/com/watabou/pixeldungeon/actors/buffs/Levitation.java

package com.watabou.pixeldungeon.actors.buffs;

import com.watabou.pixeldungeon.Dungeon;
import com.watabou.pixeldungeon.actors.Char;
import com.watabou.pixeldungeon.sprites.CharSprite;
import com.watabou.pixeldungeon.ui.BuffIndicator;

import org.jetbrains.annotations.NotNull;

public class Levitation extends FlavourBuff {

	public static final float DURATION	= 20f;
	
	@Override
	public boolean attachTo(@NotNull Char target ) {
		if (super.attachTo( target )) {
			Roots.detach( target, Roots.class );
			return true;
		} else {
			return false;
		}
	}
	
	@Override
	public void detach() {
		super.detach();
		if(Dungeon.level != null) {
			Dungeon.level.press(target.getPos(), target);
		}
	}

	@Override
	public int icon() {
		return BuffIndicator.LEVITATION;
	}

	@Override
	public CharSprite.State charSpriteStatus() {
		return CharSprite.State.LEVITATING;
	}
}

File: scripts/buffs/Levitation.lua
---
--- Levitation: floats over traps/terrain, presses cell on landing
--- (was actors/buffs/Levitation.java)
---
local RPD  = require "scripts/lib/commonClasses"
local buff = require "scripts/lib/buff"

return buff.init{
    desc  = function ()
        return {
            icon          = 1, -- BuffIndicator.LEVITATION
            name          = "LevitationBuff_Name",
            info          = "LevitationBuff_Info",
        }
    end,

    attached = function(self, buff)
        RPD.removeBuff(buff.target, "Roots")
    end,

    detach = function(self, buff)
        local level = RPD.Dungeon.level
        if level then
            level:press(buff.target:getPos(), buff.target)
        end
    end,

    charSpriteStatus = function(self, buff)
        return "LEVITATING"
    end
}
